### PR TITLE
Save and display per-token attention maps

### PR DESCRIPTION
**Pull request by damian0815**
This pull request enables the display of per-token attention maps after generating an image.

### Done: 
* [x] Collect and return attention maps to generate.py
* [x] Pass attention maps and tokens to webUI - currently pushed as the following fields on the object emitted to the webUI socket with `generationResult`:
  - `attentionMaps` (base64 image, size `width/8 x 77*height/8`), and
  - `tokens`, see below.

### Todo:
* [ ] Display maps - @psychedelicious and/or @blessedcoolant will need your help on this part

Typical content of the `tokens` array is, eg for prompt `a fluffy miyazaki dog`, an array: `['a</w>', 'fluffy</w>', 'mi', 'yaz', 'aki</w>', 'dog']`. The `</w>` strings represent "end-of-word". With this implementation, to match tokens to fragments of the input prompt text in the input box, the frontend code is going to have to crawl through the prompt and do a best-fit match of these tokens to the prompt string.

### CODE REVIEW

**Comment by blessedcoolant** [comment]
Generation error:  `AttributeError: 'CrossAttention' object has no attribute 'cached_mem_free_total'`

**Comment by damian0815** [comment]
this should be merged into main asap, even if that means frontend isn't using it. 

attention map collection is always-on in this PR but the memory/performance impact is negligible.

File: backend/invoke_ai_web_server.py [hunk 1]
```diff
@@ -1564,6 +1571,19 @@ def dataURL_to_image(dataURL: str) -> ImageType:
...
+
+def image_to_dataURL(image):
+    buffered = io.BytesIO()
+    image.save(buffered, format="PNG")
+    image_base64 = "data:image/png;base64," + base64.b64encode(
+        buffered.getvalue()
+    ).decode("UTF-8")
+    return image_base64
+
+
```
Comment (psychedelicious) [inline backend/invoke_ai_web_server.py:1586]:
```suggestion
"""
Converts an image into a base64 image dataURL.
"""

def image_to_dataURL(image: ImageType) -> str:
    buffered = io.BytesIO()
    image.save(buffered, format="PNG")
    image_base64 = "data:image/png;base64," + base64.b64encode(
        buffered.getvalue()
    ).decode("UTF-8")
    return image_base64


```

Just adding types here for clarity (I have imported the PIL Image type as ImageType).

**Review by psychedelicious — CHANGES_REQUESTED** [review]
1 simple change requested - adding types to util function. I haven't been consistent with this at all but I want to make an effort to add types whenever possible going forward.

This is not related to this particular PR, but we will need to save the attention map data somewhere besides just the gallery. The gallery images arrays are not persisted across reloads, and even if they were, resetting localstorage will of course clear them.

So we need to think about how to store them as metadata.

**Review by psychedelicious — APPROVED** [review]

**Pull request merged**